src/WebJobs.Script/Diagnostics/DictionaryLoggerScope.cs

```diff
@@ -13,27 +14,56 @@ namespace Microsoft.Azure.WebJobs.Script.Diagnostics
     /// </summary>
     internal class DictionaryLoggerScope
     {
-        private readonly object _state;
-
         private static AsyncLocal<DictionaryLoggerScope> _value = new AsyncLocal<DictionaryLoggerScope>();
 
-        private DictionaryLoggerScope(object state, DictionaryLoggerScope parent)
+        private DictionaryLoggerScope(IReadOnlyDictionary<string, object> state, DictionaryLoggerScope parent)
         {
-            _state = state;
+            State = state;
             Parent = parent;
         }
 
+        internal IReadOnlyDictionary<string, object> State { get; private set; }
+
         internal DictionaryLoggerScope Parent { get; private set; }
 
         public static DictionaryLoggerScope Current
         {
-            get => _value.Value;
-            set => _value.Value = value;
+            get
+            {
+                return _value.Value;
+            }
+
+            set
+            {
+                _value.Value = value;
+            }
         }
 
         public static IDisposable Push(object state)
         {
-            Current = new DictionaryLoggerScope(state, Current);
+            var stateValues = state as IDictionary<string, object>;
+            if (stateValues == null)
+            {
+                var stateEnum = state as IEnumerable<KeyValuePair<string, object>>;
+                if (stateEnum != null)
+                {
+                    // Convert this to a dictionary as we have scenarios where we cannot have duplicates. In this
+                    // case, if there are dupes, the later entry wins.
+                    stateValues = new Dictionary<string, object>();
+                    foreach (var entry in stateEnum)
+                    {
+                        stateValues[entry.Key] = entry.Value;
+                    }
+                }
+            }
+
+            if (stateValues == null)
+            {
+                // There's nothing we can do with other states.
```

Comment:
Sometimes you just can't do anything with a specific state. For example, the Console logger can take a string here and it'll indent your logs with that label. But for us, I'm not sure what to do with a string (or an int, or a custom object). Other Loggers may be able to handle them, but we'll just ignore it.